File: src/forecast_warden/incidents.py
"""Incident creation, persistence (md + json), and resolution."""

from __future__ import annotations

import hashlib
import json
import re
from datetime import datetime, timezone
from pathlib import Path

from forecast_warden.models import (
    Finding,
    Hypothesis,
    Incident,
    IncidentStatus,
    Severity,
    SuggestedAction,
)

SEVERITY_RANK = {Severity.INFO: 0, Severity.WARNING: 1, Severity.CRITICAL: 2}


def actionable_findings(findings: list[Finding]) -> list[Finding]:
    """Findings that can open an incident (warning or critical)."""
    return [f for f in findings if f.severity in (Severity.WARNING, Severity.CRITICAL)]


def max_severity(findings: list[Finding]) -> Severity:
    if not findings:
        return Severity.INFO
    return max(findings, key=lambda f: SEVERITY_RANK[f.severity]).severity


def suggest_hypotheses(findings: list[Finding]) -> list[Hypothesis]:
    codes = {f.code for f in findings}
    hyps: list[Hypothesis] = []
    if "HighMAPE" in codes:
        hyps.extend([Hypothesis.DEMAND_SPIKE, Hypothesis.MODEL_STALE])
    if "BiasShift" in codes:
        hyps.extend([Hypothesis.FEATURE_BREAK, Hypothesis.CALENDAR])
    if "CoverageBreak" in codes:
        hyps.append(Hypothesis.MODEL_STALE)
    if "LowSupport" in codes:
        hyps.append(Hypothesis.DATA_DELAY)
    # unique preserve order
    seen: set[Hypothesis] = set()
    out: list[Hypothesis] = []
    for h in hyps:
        if h not in seen:
            seen.add(h)
            out.append(h)
    return out or [Hypothesis.MODEL_STALE]


def suggest_action(severity: Severity, findings: list[Finding]) -> SuggestedAction:
    codes = {f.code for f in findings}
    if severity == Severity.CRITICAL:
        return SuggestedAction.HOLD_PROMOTION
    if "BiasShift" in codes and "HighMAPE" in codes:
        return SuggestedAction.RETRAIN
    return SuggestedAction.INVESTIGATE


def make_incident_id(run_id: str, findings: list[Finding]) -> str:
    compact = run_id.replace("-", "")
    payload = "|".join(
        sorted(f"{f.code}:{f.entity_id}:{f.severity.value}" for f in findings)
    )
    short = hashlib.sha1(payload.encode()).hexdigest()[:4]
    return f"fw-{compact}-{short}"


def build_incident(run_id: str, findings: list[Finding], created_at: datetime | None = None) -> Incident | None:
    action = actionable_findings(findings)
    if not action:
        return None
    created = created_at or datetime.now(timezone.utc)
    severity = max_severity(action)
    entities = sorted({f.entity_id for f in action})
    codes = sorted({f.code for f in action})
    return Incident(
        id=make_incident_id(run_id, action),
        run_id=run_id,
        status=IncidentStatus.OPEN,
        severity=severity,
        entities=entities,
        detector_codes=codes,
        created_at=created,
        findings=findings,  # keep info findings too for context
        hypotheses=suggest_hypotheses(findings),
        suggested_action=suggest_action(severity, action),
    )


def incident_paths(incidents_dir: Path, incident: Incident) -> tuple[Path, Path]:
    short = incident.id.split("-")[-1]
    stem = f"{incident.run_id}__{short}"
    return incidents_dir / f"{stem}.md", incidents_dir / f"{stem}.json"


def write_incident(incidents_dir: Path, incident: Incident) -> tuple[Path, Path]:
    incidents_dir.mkdir(parents=True, exist_ok=True)
    md_path, json_path = incident_paths(incidents_dir, incident)
    json_path.write_text(
        incident.model_dump_json(indent=2),
        encoding="utf-8",
    )
    md_path.write_text(render_markdown(incident), encoding="utf-8")
    return md_path, json_path


def render_markdown(incident: Incident) -> str:
    entities = ", ".join(f"`{e}`" for e in incident.entities)
    codes = ", ".join(incident.detector_codes)
    hyp_lines = "\n".join(f"- `{h.value}`" for h in incident.hypotheses)
    rows = []
    for f in incident.findings:
        ev = ", ".join(f"{k}={v}" for k, v in f.evidence.items())
        rows.append(
            f"| {f.code} | {f.severity.value} | {f.entity_id} | {ev} | {f.hint} |"
        )
    table = "\n".join(rows) if rows else "| — | — | — | — | — |"
    gate = (
        "Model promotion / config change blocked while status is `open` "
        "(resolve or accept-risk first)."
        if incident.severity in (Severity.WARNING, Severity.CRITICAL)
        else "No gate."
    )
    note_block = f"\n**Resolution note:** {incident.note}\n" if incident.note else ""
    return f"""---
id: {incident.id}
run_id: "{incident.run_id}"
status: {incident.status.value}
severity: {incident.severity.value}
entities: {incident.entities}
detector_codes: {incident.detector_codes}
created_at: {incident.created_at.isoformat()}
suggested_action: {incident.suggested_action.value}
---

# Forecast Incident `{incident.id}`

## Symptom

Run `{incident.run_id}` flagged {entities} via {codes}
(severity **{incident.severity.value}**). Review evidence before promoting the model.
{note_block}
## Evidence

| detector | severity | entity | evidence | hint |
|----------|----------|--------|----------|------|
{table}

## Hypotheses (candidate)

{hyp_lines}

## Suggested action

`{incident.suggested_action.value}`

## Gate

{gate}
"""


def list_incidents(
    incidents_dir: Path,
    status: IncidentStatus | None = None,
) -> list[Incident]:
    if not incidents_dir.exists():
        return []
    incidents: list[Incident] = []
    for path in sorted(incidents_dir.glob("*.json")):
        data = json.loads(path.read_text(encoding="utf-8"))
        inc = Incident.model_validate(data)
        if status is None or inc.status == status:
            incidents.append(inc)
    return incidents


def load_incident(incidents_dir: Path, incident_id: str) -> Incident | None:
    for inc in list_incidents(incidents_dir):
        if inc.id == incident_id:
            return inc
    # also allow short suffix / stem match
    for path in incidents_dir.glob("*.json") if incidents_dir.exists() else []:
        data = json.loads(path.read_text(encoding="utf-8"))
        inc = Incident.model_validate(data)
        if incident_id in (inc.id, path.stem, path.stem.split("__")[-1]):
            return inc
    return None


def resolve_incident(
    incidents_dir: Path,
    incident_id: str,
    note: str,
    status: IncidentStatus = IncidentStatus.RESOLVED,
) -> Incident | None:
    inc = load_incident(incidents_dir, incident_id)
    if inc is None:
        return None
    updated = inc.model_copy(
        update={
            "status": status,
            "note": note,
            "resolved_at": datetime.now(timezone.utc),
        }
    )
    write_incident(incidents_dir, updated)
    return updated


FRONTMATTER_RE = re.compile(r"^---\n(.*?)\n---\n", re.DOTALL)

File: internal/incidents/incidents.go
package incidents

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/luizpedrini/forecast-warden/internal/models"
)

func ActionableFindings(findings []models.Finding) []models.Finding {
	var out []models.Finding
	for _, f := range findings {
		if f.Severity == models.SeverityWarning || f.Severity == models.SeverityCritical {
			out = append(out, f)
		}
	}
	return out
}

func MaxSeverity(findings []models.Finding) models.Severity {
	if len(findings) == 0 {
		return models.SeverityInfo
	}
	max := findings[0].Severity
	for _, f := range findings[1:] {
		if f.Severity.Rank() > max.Rank() {
			max = f.Severity
		}
	}
	return max
}

func SuggestHypotheses(findings []models.Finding) []models.Hypothesis {
	codes := map[string]struct{}{}
	for _, f := range findings {
		codes[f.Code] = struct{}{}
	}
	var hyps []models.Hypothesis
	if _, ok := codes["HighMAPE"]; ok {
		hyps = append(hyps, models.HypDemandSpike, models.HypModelStale)
	}
	if _, ok := codes["HighWAPE"]; ok {
		hyps = append(hyps, models.HypDemandSpike, models.HypModelStale)
	}
	if _, ok := codes["HighRMSE"]; ok {
		hyps = append(hyps, models.HypModelStale)
	}
	if _, ok := codes["BiasShift"]; ok {
		hyps = append(hyps, models.HypFeatureBreak, models.HypCalendar)
	}
	if _, ok := codes["CoverageBreak"]; ok {
		hyps = append(hyps, models.HypModelStale)
	}
	if _, ok := codes["UnstableForecast"]; ok {
		hyps = append(hyps, models.HypModelStale, models.HypFeatureBreak)
	}
	if _, ok := codes["LowSupport"]; ok {
		hyps = append(hyps, models.HypDataDelay)
	}
	seen := map[models.Hypothesis]struct{}{}
	var out []models.Hypothesis
	for _, h := range hyps {
		if _, ok := seen[h]; ok {
			continue
		}
		seen[h] = struct{}{}
		out = append(out, h)
	}
	if len(out) == 0 {
		return []models.Hypothesis{models.HypModelStale}
	}
	return out
}

func SuggestAction(severity models.Severity, findings []models.Finding) models.SuggestedAction {
	codes := map[string]struct{}{}
	for _, f := range findings {
		codes[f.Code] = struct{}{}
	}
	if severity == models.SeverityCritical {
		return models.ActionHoldPromotion
	}
	_, bias := codes["BiasShift"]
	_, mape := codes["HighMAPE"]
	_, wape := codes["HighWAPE"]
	if bias && (mape || wape) {
		return models.ActionRetrain
	}
	return models.ActionInvestigate
}

// SanitizeRunID normalizes a run_id for use in incident IDs and file paths.
// It keeps alphanumerics, hyphens, and underscores; replaces other chars with underscore.
func SanitizeRunID(runID string) string {
	var sb strings.Builder
	sb.Grow(len(runID))
	for _, r := range runID {
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') || r == '-' || r == '_' {
			sb.WriteRune(r)
		} else {
			sb.WriteRune('_')
		}
	}
	return sb.String()
}

// MakeIncidentID returns a stable incident ID based solely on run_id.
// Format: fw-{sanitized_run_id}
// This ensures the same run always maps to the same incident, allowing in-place updates.
func MakeIncidentID(runID string) string {
	return fmt.Sprintf("fw-%s", SanitizeRunID(runID))
}

func BuildIncident(runID string, findings []models.Finding, createdAt time.Time) *models.Incident {
	action := ActionableFindings(findings)
	if len(action) == 0 {
		return nil
	}
	if createdAt.IsZero() {
		createdAt = time.Now().UTC()
	}
	severity := MaxSeverity(action)
	entitySet := map[string]struct{}{}
	codeSet := map[string]struct{}{}
	for _, f := range action {
		entitySet[f.EntityID] = struct{}{}
		codeSet[f.Code] = struct{}{}
	}
	entities := make([]string, 0, len(entitySet))
	for e := range entitySet {
		entities = append(entities, e)
	}
	sort.Strings(entities)
	codes := make([]string, 0, len(codeSet))
	for c := range codeSet {
		codes = append(codes, c)
	}
	sort.Strings(codes)
	return &models.Incident{
		ID:              MakeIncidentID(runID),
		RunID:           runID,
		Status:          models.StatusOpen,
		Severity:        severity,
		Entities:        entities,
		DetectorCodes:   codes,
		CreatedAt:       createdAt,
		Findings:        findings,
		Hypotheses:      SuggestHypotheses(findings),
		SuggestedAction: SuggestAction(severity, action),
		Note:            nil,
		ResolvedAt:      nil,
	}
}

func IncidentPaths(incidentsDir string, incident *models.Incident) (mdPath, jsonPath string) {
	// Use incident ID directly as stem (already includes sanitized run_id)
	stem := incident.ID
	return filepath.Join(incidentsDir, stem+".md"), filepath.Join(incidentsDir, stem+".json")
}

func WriteIncident(incidentsDir string, incident *models.Incident) (mdPath, jsonPath string, err error) {
	if err := os.MkdirAll(incidentsDir, 0o755); err != nil {
		return "", "", err
	}
	mdPath, jsonPath = IncidentPaths(incidentsDir, incident)
	data, err := json.MarshalIndent(incident, "", "  ")
	if err != nil {
		return "", "", err
	}
	data = append(data, '\n')
	if err := os.WriteFile(jsonPath, data, 0o644); err != nil {
		return "", "", err
	}
	if err := os.WriteFile(mdPath, []byte(RenderMarkdown(incident)), 0o644); err != nil {
		return "", "", err
	}
	return mdPath, jsonPath, nil
}

func RenderMarkdown(incident *models.Incident) string {
	ents := make([]string, len(incident.Entities))
	for i, e := range incident.Entities {
		ents[i] = fmt.Sprintf("`%s`", e)
	}
	entities := strings.Join(ents, ", ")
	codes := strings.Join(incident.DetectorCodes, ", ")
	hypLines := make([]string, 0, len(incident.Hypotheses))
	for _, h := range incident.Hypotheses {
		hypLines = append(hypLines, fmt.Sprintf("- `%s`", h))
	}
	rows := make([]string, 0, len(incident.Findings))
	for _, f := range incident.Findings {
		evParts := make([]string, 0, len(f.Evidence))
		keys := make([]string, 0, len(f.Evidence))
		for k := range f.Evidence {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			evParts = append(evParts, fmt.Sprintf("%s=%v", k, f.Evidence[k]))
		}
		rows = append(rows, fmt.Sprintf("| %s | %s | %s | %s | %s |",
			f.Code, f.Severity, f.EntityID, strings.Join(evParts, ", "), f.Hint))
	}
	table := strings.Join(rows, "\n")
	if table == "" {
		table = "| — | — | — | — | — |"
	}
	gate := "No gate."
	if incident.Severity == models.SeverityWarning || incident.Severity == models.SeverityCritical {
		gate = "Model promotion / config change blocked while status is `open` (resolve or accept-risk first)."
	}
	noteBlock := ""
	if incident.Note != nil && *incident.Note != "" {
		noteBlock = fmt.Sprintf("\n**Resolution note:** %s\n", *incident.Note)
	}
	entitiesYAML, _ := json.Marshal(incident.Entities)
	codesYAML, _ := json.Marshal(incident.DetectorCodes)
	return fmt.Sprintf(`---
id: %s
run_id: "%s"
status: %s
severity: %s
entities: %s
detector_codes: %s
created_at: %s
suggested_action: %s
---

# Forecast Incident `+"`%s`"+`

## Symptom

Run `+"`%s`"+` flagged %s via %s
(severity **%s**). Review evidence before promoting the model.
%s
## Evidence

| detector | severity | entity | evidence | hint |
|----------|----------|--------|----------|------|
%s

## Hypotheses (candidate)

%s

## Suggested action

`+"`%s`"+`

## Gate

%s
`,
		incident.ID,
		incident.RunID,
		incident.Status,
		incident.Severity,
		string(entitiesYAML),
		string(codesYAML),
		incident.CreatedAt.Format(time.RFC3339Nano),
		incident.SuggestedAction,
		incident.ID,
		incident.RunID,
		entities,
		codes,
		incident.Severity,
		noteBlock,
		table,
		strings.Join(hypLines, "\n"),
		incident.SuggestedAction,
		gate,
	)
}

func ListIncidents(incidentsDir string, status *models.IncidentStatus) ([]models.Incident, error) {
	entries, err := filepath.Glob(filepath.Join(incidentsDir, "*.json"))
	if err != nil {
		return nil, err
	}
	sort.Strings(entries)
	var out []models.Incident
	for _, path := range entries {
		data, err := os.ReadFile(path)
		if err != nil {
			return nil, err
		}
		var inc models.Incident
		if err := json.Unmarshal(data, &inc); err != nil {
			return nil, err
		}
		if status != nil && inc.Status != *status {
			continue
		}
		out = append(out, inc)
	}
	return out, nil
}

func LoadIncident(incidentsDir, incidentID string) (*models.Incident, error) {
	incs, err := ListIncidents(incidentsDir, nil)
	if err != nil {
		return nil, err
	}
	for i := range incs {
		if incs[i].ID == incidentID {
			return &incs[i], nil
		}
	}
	entries, err := filepath.Glob(filepath.Join(incidentsDir, "*.json"))
	if err != nil {
		return nil, err
	}
	for _, path := range entries {
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		var inc models.Incident
		if err := json.Unmarshal(data, &inc); err != nil {
			continue
		}
		stem := strings.TrimSuffix(filepath.Base(path), ".json")
		short := ""
		if parts := strings.Split(stem, "__"); len(parts) > 1 {
			short = parts[len(parts)-1]
		}
		if incidentID == inc.ID || incidentID == stem || incidentID == short {
			return &inc, nil
		}
	}
	return nil, nil
}

func ResolveIncident(incidentsDir, incidentID, note string, status models.IncidentStatus) (*models.Incident, error) {
	inc, err := LoadIncident(incidentsDir, incidentID)
	if err != nil {
		return nil, err
	}
	if inc == nil {
		return nil, nil
	}
	now := time.Now().UTC()
	inc.Status = status
	inc.Note = &note
	inc.ResolvedAt = &now
	if _, _, err := WriteIncident(incidentsDir, inc); err != nil {
		return nil, err
	}
	return inc, nil
}

// UpdateIncidentInPlace updates an existing open incident with new findings.
// Preserves: createdAt, history. Updates: findings, severity, entities, codes, hypotheses, action.
func UpdateIncidentInPlace(existing, new *models.Incident) *models.Incident {
	now := time.Now().UTC()

	// Start with the new incident data but preserve identity and history
	result := *new
	result.CreatedAt = existing.CreatedAt
	result.Note = existing.Note
	result.ResolvedAt = existing.ResolvedAt

	// Preserve and extend history
	result.History = append([]models.HistoryEvent(nil), existing.History...)
	result.History = append(result.History, models.HistoryEvent{
		Timestamp: now,
		Event:     "updated",
		Status:    result.Status,
		Severity:  result.Severity,
	})

	return &result
}

// ReopenIncident reopens a resolved incident with new critical findings.
// Preserves: createdAt, full history including resolution. Updates: status to open, new findings/severity.
func ReopenIncident(existing, new *models.Incident) *models.Incident {
	now := time.Now().UTC()

	// Start with the new incident data but preserve identity and history
	result := *new
	result.CreatedAt = existing.CreatedAt
	result.Status = models.StatusOpen

	// Preserve and extend history, including the prior resolution
	result.History = append([]models.HistoryEvent(nil), existing.History...)

	// Record the reopen event with context from the old resolution
	oldNote := ""
	if existing.Note != nil {
		oldNote = *existing.Note
	}
	result.History = append(result.History, models.HistoryEvent{
		Timestamp: now,
		Event:     "reopened",
		Status:    result.Status,
		Severity:  result.Severity,
		Note:      fmt.Sprintf("reopened from %s (was: %s)", existing.Status, oldNote),
	})

	// Clear the resolved state but keep notes in history
	result.Note = nil
	result.ResolvedAt = nil

	return &result
}
